fix(typeops): do not copy non-existent type fields

saem opened the pull request:
## Summary

During liftdestructors, do not produce hidden type field copies for
types without such fields, preventing invalid code from being generated.

## Details

`liftdestructors` produces symbols as it proceeds via `produceSym`. This
procedure not only produces a symbol, but ensures it's initialized,
type operations ("destructors") are attached, and necessary ops called.
One such operation is copying of the hidden type field for objects types
that have them.

This latter opertation was guarded by `isObjLackingTypeField` that
expects any type passed to it be unwrapped (i.e. removing `tyDistinct`,
`tyAlias`, etc). That wasn't the case and so `liftdestructors` was
producing copy operations for hidden type fields that didn't exist. Now
the type passed to `isObjLackingTypeField` is first unwrapped.

A test, `ttypeops_misc`, has also been added for this (contributed by
@zerbina).

saem committed:
fix(typeops): do not copy non-existent type fields

## Summary

During liftdestructors, do not produce hidden type field copies for
types without such fields, preventing invalid code from being generated.

## Details

`liftdestructors` produces symbols as it proceeds via `produceSym`. This
procedure not only produces a symbol, but ensures it's initialized,
type operations ("destructors") are attached, and necessary ops called.
One such operation is copying of the hidden type field for objects types
that have them.

This latter opertation was guarded by `isObjLackingTypeField` that
expects any type passed to it be unwrapped (i.e. removing `tyDistinct`,
`tyAlias`, etc). That wasn't the case and so `liftdestructors` was
producing copy operations for hidden type fields that didn't exist. Now
the type passed to `isObjLackingTypeField` is first unwrapped.

A test, `ttypeops_misc`, has also been added for this (contributed by
@zerbina).

zerbina commented on tests/lang_objects/destructor/ttypeops_misc.nim:
```diff
@@ -0,0 +1,18 @@
+  ## previously this led to a code gen bug where a non-existent hidden type
+  ## field was being copied. the various wrappers alias/distinct weren't being
+  ## skipped when testing to see if the underlying type was an object.
+  type
+    Object = object
+      x: string # force `Object` to require lifetime hooks
```
```suggestion
  type WithHooks = object

  proc `=destroy`(x: var WithHooks) =
    # force `WithHooks` to require lifetime hooks
    discard

  type
    Object = object
      x: WithHooks # make sure that `Object` gets a synthesized copy hook
```
A soft suggestion for making the test-case independent of `--gc:arc|orc`, as only with the latter two does the `string` type have hooks.

saem replied:
good call, independence would be much better.

zerbina reviewed (approved):
Very nice! I left a single, non-blocking suggestion regarding the test.

saem committed:
make tests gc independent

define a type hook to force a definition, rather than relying on the gc being used to coerce it into happening.

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

zerbina commented:
I've removed the PR title from the PR message, as it looked like it's there by accident.

saem commented:
> I've removed the PR title from the PR message, as it looked like it's there by accident.

thanks, that was an accident. 😅

saem commented:
/merge
